src/control/cmd/dmg/pretty/firmware.go

```diff
@@ -27,34 +27,238 @@ package pretty
+	scmNotFound      = "No SCM devices detected"
+	errorPrefix      = "Error"
+)
+
+// hostDeviceSet represents a collection of hosts and devices on those hosts.
+type hostDeviceSet struct {
+	Hosts   *hostlist.HostSet
+	Devices []string
+}
+
+// AddHost adds a host to the set.
+func (h *hostDeviceSet) AddHost(host string) {
+	h.Hosts.Insert(host)
+}
+
+// AddDevice adds a device to the set.
+func (h *hostDeviceSet) AddDevice(device string) {
+	h.Devices = append(h.Devices, device)
+}
+
+// newHostDeviceSet creates and initializes an empty hostDeviceSet.
+func newHostDeviceSet() (*hostDeviceSet, error) {
+	hosts, err := hostlist.CreateSet("")
+	if err != nil {
+		return nil, err
+	}
+	return &hostDeviceSet{
+		Hosts: hosts,
+	}, nil
+}
+
+// hostDeviceResultMap is a map from a result string to a hostDeviceSet.
+type hostDeviceResultMap map[string]*hostDeviceSet
+
+// AddHostDevice adds a host and a device to the map set for a given result string.
+func (m hostDeviceResultMap) AddHostDevice(resultStr string, host string, device string) error {
+	err := m.AddHost(resultStr, host)
+	if err != nil {
+		return err
+	}
+	m[resultStr].AddDevice(device)
+	return nil
+}
+
+// AddHost adds a host to the map set for a given result string.
+func (m hostDeviceResultMap) AddHost(resultStr string, host string) error {
+	if _, ok := m[resultStr]; !ok {
+		newSet, err := newHostDeviceSet()
+		if err != nil {
+			return err
+		}
+		m[resultStr] = newSet
+	}
+
+	m[resultStr].AddHost(host)
+	return nil
+}
+
+// Keys returns the sorted keys of the map.
+func (m hostDeviceResultMap) Keys() []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+type hostDeviceError struct {
+	Host  string
+	Error error
+	DevID string
+}
+
+// PrintSCMFirmwareQueryMap formats the firmware query results in a condensed format.
+func PrintSCMFirmwareQueryMap(fwMap control.HostSCMQueryMap, out io.Writer,
+	opts ...control.PrintConfigOption) error {
+	successes, errors, err := condenseSCMQueryMap(fwMap)
+	if err != nil {
+		return err
+	}
+
+	if err = printDeviceErrorTable(errors, out, opts...); err != nil {
+		return err
+	}
+
+	return printCondensedQuerySuccesses(successes, out, opts...)
+}
+
+func condenseSCMQueryMap(fwMap control.HostSCMQueryMap) (hostDeviceResultMap, []hostDeviceError, error) {
+	successes := make(hostDeviceResultMap)
+	errors := make([]hostDeviceError, 0)
+	for _, host := range fwMap.Keys() {
+		results := fwMap[host]
+		if len(results) == 0 {
+			err := successes.AddHost(scmNotFound, host)
+			if err != nil {
+				return nil, nil, err
+			}
+			continue
+		}
+
+		for _, devRes := range results {
+			if devRes.Error == nil {
+				var b strings.Builder
+				printSCMFirmwareInfo(devRes.Info, &b)
+				err := successes.AddHostDevice(b.String(), host, devRes.Module.String())
+				if err != nil {
+					return nil, nil, err
+				}
+			} else {
```

Comment:
nit: I've been trying where possible (as advised) to avoid else and instead continue the iteration in the above if clause so you can de-dent the rest of the logic in the for loop

Reply:
Will note for the future, assuming no other changes are needed.